File: sdks/local-rp/c/src/local_rp.c
#include <stdlib.h>
#include "local_rp.h"

#include <string.h>
#include <time.h>

#include <openssl/crypto.h>

#include "error.h"
#include "time_util.h"

static const char LOCAL_RP_CALLBACK_BOX_TAG[] = "linkkeys-local-rp-callback-box";

int lrp_envelope_signature_input(const char *context, const uint8_t *payload, size_t payload_len,
                                  lrp_bytes *out, lrp_error *err) {
    cbor_buf b;
    cbor_buf_init(&b);
    int rc = 0;
    rc |= cbor_write_array_header(&b, 2);
    rc |= cbor_write_text_cstr(&b, context);
    rc |= cbor_write_bytes(&b, payload, payload_len);
    if (rc != 0) {
        cbor_buf_free(&b);
        return lrp_fail(err, LRP_ERR_OUT_OF_MEMORY, "envelope signature input: out of memory");
    }
    *out = cbor_buf_release(&b);
    return 0;
}

int lrp_sign_envelope(const char *context, const uint8_t *payload, size_t payload_len,
                       const uint8_t signing_private_key[32], lrp_bytes *out_signature,
                       lrp_error *err) {
    lrp_bytes sig_input = {0};
    if (lrp_envelope_signature_input(context, payload, payload_len, &sig_input, err) != 0) {
        return -1;
    }
    uint8_t sig[64];
    int rc = lrp_ed25519_sign(signing_private_key, sig_input.data, sig_input.len, sig, err);
    lrp_bytes_free(&sig_input);
    if (rc != 0) return -1;
    uint8_t *heap_sig = (uint8_t *)malloc(64);
    if (heap_sig == NULL) return lrp_fail(err, LRP_ERR_OUT_OF_MEMORY, "out of memory");
    memcpy(heap_sig, sig, 64);
    out_signature->data = heap_sig;
    out_signature->len = 64;
    return 0;
}

int64_t lrp_wall_clock_now(void) { return (int64_t)time(NULL); }

int lrp_check_timestamps(const char *issued_at, const char *expires_at, int64_t now_unix,
                          int64_t skew_seconds, lrp_error *err) {
    int64_t issued, expires;
    if (lrp_parse_rfc3339(issued_at, &issued, err) != 0) return -1;
    if (lrp_parse_rfc3339(expires_at, &expires, err) != 0) return -1;
    if (now_unix + skew_seconds < issued) {
        return lrp_fail(err, LRP_ERR_VERIFICATION, "timestamp is not yet valid");
    }
    if (now_unix - skew_seconds > expires) {
        return lrp_fail(err, LRP_ERR_VERIFICATION, "timestamp has expired");
    }
    return 0;
}

int lrp_check_expirations_impl(const char *expires_at, int64_t now_unix,
                                lrp_expiration_status *out, lrp_error *err) {
    int64_t expires;
    if (lrp_parse_rfc3339(expires_at, &expires, err) != 0) return -1;
    int64_t remaining = expires - now_unix;
    lrp_expiration_level level;
    if (now_unix >= expires) {
        level = LRP_EXPIRATION_EXPIRED;
    } else if (remaining <= (int64_t)30 * 86400) {
        level = LRP_EXPIRATION_CRITICAL;
    } else if (remaining <= (int64_t)90 * 86400) {
        level = LRP_EXPIRATION_WARNING;
    } else if (remaining <= (int64_t)180 * 86400) {
        level = LRP_EXPIRATION_NOTICE;
    } else {
        level = LRP_EXPIRATION_OK;
    }
    out->level = level;
    out->expires_at_unix = expires;
    out->now_unix = now_unix;
    return 0;
}

/* Constant-time byte-equality (SEC fix): the length check is inherently
 * public (nonce/state are always caller-generated, fixed-size values, so
 * leaking their length is not a secret), but the byte comparison itself
 * must not branch on secret content — `CRYPTO_memcmp` (not `memcmp`, which
 * OpenSSL and most libc implementations short-circuit on the first
 * mismatching byte) is OpenSSL's documented constant-time comparison
 * primitive. */
static int constant_time_bytes_eq(const uint8_t *a, size_t a_len, const uint8_t *b, size_t b_len) {
    if (a_len != b_len) return 0;
    if (a_len == 0) return 1;
    return CRYPTO_memcmp(a, b, a_len) == 0;
}

int lrp_verify_nonce_state(const uint8_t *expected_nonce, size_t expected_nonce_len,
                            const uint8_t *expected_state, size_t expected_state_len,
                            const uint8_t *actual_nonce, size_t actual_nonce_len,
                            const uint8_t *actual_state, size_t actual_state_len, lrp_error *err) {
    if (!constant_time_bytes_eq(expected_nonce, expected_nonce_len, actual_nonce, actual_nonce_len)) {
        return lrp_fail(err, LRP_ERR_VERIFICATION, "nonce does not match");
    }
    if (!constant_time_bytes_eq(expected_state, expected_state_len, actual_state, actual_state_len)) {
        return lrp_fail(err, LRP_ERR_VERIFICATION, "state does not match");
    }
    return 0;
}

int lrp_verify_audience(const char *payload_audience_fingerprint, const char *local_rp_fingerprint,
                         lrp_error *err) {
    if (strcmp(payload_audience_fingerprint, local_rp_fingerprint) != 0) {
        return lrp_fail(err, LRP_ERR_VERIFICATION, "audience fingerprint does not match");
    }
    return 0;
}

int lrp_verify_issuer(const char *payload_user_domain, const char *expected_domain,
                       lrp_error *err) {
    if (strcmp(payload_user_domain, expected_domain) != 0) {
        return lrp_fail(err, LRP_ERR_VERIFICATION, "issuing domain does not match");
    }
    return 0;
}

int lrp_verify_callback_url(const char *payload_callback_url, const char *arrived_url,
                             lrp_error *err) {
    if (strcmp(payload_callback_url, arrived_url) != 0) {
        return lrp_fail(err, LRP_ERR_VERIFICATION, "callback URL does not match");
    }
    return 0;
}

static int bytes_eq(const lrp_bytes *a, const uint8_t *b, size_t b_len) {
    return a->len == b_len && (b_len == 0 || memcmp(a->data, b, b_len) == 0);
}

int lrp_check_callback_header_matches_payload(const lrp_w_callback_header *header,
                                               const lrp_w_callback_payload *payload,
                                               lrp_error *err) {
    if (strcmp(header->fingerprint.data, payload->audience_fingerprint.data) != 0) {
        return lrp_fail(err, LRP_ERR_VERIFICATION,
                         "callback header does not match signed payload field: fingerprint");
    }
    if (!bytes_eq(&payload->nonce, header->nonce.data, header->nonce.len)) {
        return lrp_fail(err, LRP_ERR_VERIFICATION,
                         "callback header does not match signed payload field: nonce");
    }
    if (!bytes_eq(&payload->state, header->state.data, header->state.len)) {
        return lrp_fail(err, LRP_ERR_VERIFICATION,
                         "callback header does not match signed payload field: state");
    }
    if (strcmp(header->issued_at.data, payload->issued_at.data) != 0) {
        return lrp_fail(err, LRP_ERR_VERIFICATION,
                         "callback header does not match signed payload field: issued_at");
    }
    if (strcmp(header->expires_at.data, payload->expires_at.data) != 0) {
        return lrp_fail(err, LRP_ERR_VERIFICATION,
                         "callback header does not match signed payload field: expires_at");
    }
    return 0;
}

int lrp_check_signing_key_valid(const lrp_domain_public_key *key, lrp_error *err) {
    if (strcmp(key->key_usage.data, "sign") != 0) {
        return lrp_fail(err, LRP_ERR_VERIFICATION, "signature verification failed");
    }
    if (key->revoked_at.data != NULL) {
        return lrp_fail(err, LRP_ERR_VERIFICATION, "signing key has been revoked: %s",
                         key->key_id.data);
    }
    int64_t expires;
    if (lrp_parse_rfc3339(key->expires_at.data, &expires, NULL) != 0) {
        return lrp_fail(err, LRP_ERR_VERIFICATION, "signing key has an invalid expires_at: %s",
                         key->key_id.data);
    }
    if (lrp_wall_clock_now() > expires) {
        return lrp_fail(err, LRP_ERR_VERIFICATION, "signing key has expired: %s", key->key_id.data);
    }
    return 0;
}

int lrp_verify_local_rp_callback_payload(const lrp_w_signed_callback_payload *signed_payload,
                                          const lrp_domain_public_key *domain_keys,
                                          size_t domain_keys_count, int64_t now_unix,
                                          int64_t skew_seconds, lrp_w_callback_payload *out_payload,
                                          lrp_error *err) {
    const lrp_domain_public_key *key = NULL;
    for (size_t i = 0; i < domain_keys_count; i++) {
        if (strcmp(domain_keys[i].key_id.data, signed_payload->signing_key_id.data) == 0) {
            key = &domain_keys[i];
            break;
        }
    }
    if (key == NULL) {
        return lrp_fail(err, LRP_ERR_VERIFICATION, "signing key not found: %s",
                         signed_payload->signing_key_id.data);
    }
    if (lrp_check_signing_key_valid(key, err) != 0) return -1;

    lrp_bytes sig_input = {0};
    if (lrp_envelope_signature_input(LRP_CTX_CALLBACK, signed_payload->payload.data,
                                      signed_payload->payload.len, &sig_input, err) != 0) {
        return -1;
    }
    int vrc = lrp_resolve_and_verify(key->algorithm.data, sig_input.data, sig_input.len,
                                      signed_payload->signature.data, signed_payload->signature.len,
                                      key->public_key.data, key->public_key.len, err);
    lrp_bytes_free(&sig_input);
    if (vrc != 0) return -1;

    if (lrp_decode_callback_payload(signed_payload->payload.data, signed_payload->payload.len,
                                     out_payload, err) != 0) {
        return -1;
    }
    if (lrp_check_timestamps(out_payload->issued_at.data, out_payload->expires_at.data, now_unix,
                              skew_seconds, err) != 0) {
        lrp_w_callback_payload_free(out_payload);
        memset(out_payload, 0, sizeof(*out_payload));
        return -1;
    }
    return 0;
}

/* --------------------------------------------------------------------- */
/* Callback sealed box                                                   */
/* --------------------------------------------------------------------- */

int lrp_local_rp_callback_kdf(lrp_aead_suite suite, const uint8_t ephemeral_public[32],
                               const uint8_t recipient_public[32], const uint8_t *shared_secret,
                               size_t shared_secret_len, uint8_t out_key[32], lrp_bytes *out_context,
                               lrp_error *err) {
    const char *suite_id = lrp_aead_suite_str(suite);
    size_t suite_id_len = strlen(suite_id);
    size_t tag_len = sizeof(LOCAL_RP_CALLBACK_BOX_TAG) - 1;
    size_t ctx_len = tag_len + suite_id_len + 32 + 32;
    uint8_t *ctx = (uint8_t *)malloc(ctx_len);
    if (ctx == NULL) return lrp_fail(err, LRP_ERR_OUT_OF_MEMORY, "out of memory");
    size_t off = 0;
    memcpy(ctx + off, LOCAL_RP_CALLBACK_BOX_TAG, tag_len);
    off += tag_len;
    memcpy(ctx + off, suite_id, suite_id_len);
    off += suite_id_len;
    memcpy(ctx + off, ephemeral_public, 32);
    off += 32;
    memcpy(ctx + off, recipient_public, 32);
    off += 32;

    if (lrp_hkdf_sha256(shared_secret, shared_secret_len, ctx, ctx_len, out_key, err) != 0) {
        free(ctx);
        return -1;
    }
    out_context->data = ctx;
    out_context->len = ctx_len;
    return 0;
}

static int suite_in_list(const char *suite_str, const char *const *allowed, size_t allowed_count) {
    for (size_t i = 0; i < allowed_count; i++) {
        if (strcmp(suite_str, allowed[i]) == 0) return 1;
    }
    return 0;
}

int lrp_open_local_rp_callback(const uint8_t *header_bytes, size_t header_len,
                                const uint8_t *ciphertext, size_t ciphertext_len,
                                const uint8_t recipient_private_key[32],
                                const uint8_t recipient_public_key[32],
                                const char *const *allowed_suites, size_t allowed_suites_count,
                                lrp_w_callback_header *out_header,
                                lrp_w_signed_callback_payload *out_signed_payload,
                                lrp_error *err) {
    memset(out_header, 0, sizeof(*out_header));
    memset(out_signed_payload, 0, sizeof(*out_signed_payload));

    if (lrp_decode_callback_header(header_bytes, header_len, out_header, err) != 0) return -1;

    lrp_aead_suite suite;
    if (lrp_aead_suite_parse(out_header->suite.data, &suite) != 0) {
        lrp_fail(err, LRP_ERR_VERIFICATION, "unsupported AEAD suite: %s", out_header->suite.data);
        goto fail_header;
    }
    if (!suite_in_list(out_header->suite.data, allowed_suites, allowed_suites_count)) {
        lrp_fail(err, LRP_ERR_VERIFICATION, "AEAD suite was not advertised/allowed: %s",
                 out_header->suite.data);
        goto fail_header;
    }

    uint8_t shared_secret[32];
    if (lrp_x25519_ecdh(recipient_private_key, out_header->ephemeral_public_key, shared_secret,
                         err) != 0) {
        goto fail_header;
    }

    uint8_t aead_key[32];
    lrp_bytes kdf_context = {0};
    if (lrp_local_rp_callback_kdf(suite, out_header->ephemeral_public_key, recipient_public_key,
                                   shared_secret, 32, aead_key, &kdf_context, err) != 0) {
        goto fail_header;
    }

    uint8_t *aad = (uint8_t *)malloc(kdf_context.len + header_len);
    if (aad == NULL) {
        lrp_bytes_free(&kdf_context);
        lrp_fail(err, LRP_ERR_OUT_OF_MEMORY, "out of memory");
        goto fail_header;
    }
    memcpy(aad, kdf_context.data, kdf_context.len);
    memcpy(aad + kdf_context.len, header_bytes, header_len);
    size_t aad_len = kdf_context.len + header_len;
    lrp_bytes_free(&kdf_context);

    lrp_bytes plaintext = {0};
    int drc = lrp_aead_decrypt(suite, aead_key, out_header->aead_nonce, aad, aad_len, ciphertext,
                                ciphertext_len, &plaintext, err);
    free(aad);
    if (drc != 0) goto fail_header;

    int decrc =
        lrp_decode_signed_callback_payload(plaintext.data, plaintext.len, out_signed_payload, err);
    lrp_bytes_free(&plaintext);
    if (decrc != 0) goto fail_header;

    return 0;

fail_header:
    lrp_w_callback_header_free(out_header);
    memset(out_header, 0, sizeof(*out_header));
    return -1;
}

File: sdks/local-rp/ocaml/lib/local_rp.ml
(* DNS-less local RP identity: pure protocol helpers.

   Mirrors [crates/liblinkkeys/src/local_rp.rs] and the Ruby port's
   [local_rp.rb] (read that file's module docs and
   dns-less-local-rp-design.md's "Wire Precision (Normative)" section first
   -- this module implements it byte-for-byte). Summary of the shape:

   - Every signed structure uses the envelope pattern: the payload is
     CBOR-encoded once, and the signature covers
     [CBOR([context: tstr, payload: bstr])] -- a two-element CBOR array,
     never a bare [context || payload] concatenation (see
     [envelope_signature_input]).
   - Four mandatory, structure-specific context strings stop a signature
     over one structure from ever verifying as another.
   - The descriptor, login request, and ticket-redemption envelopes are
     self-asserted (verified against the local RP's own embedded signing
     key, SSH-host style). The callback payload envelope is domain-signed
     (verified against fetched domain public keys, keyed by
     [signing_key_id]).
   - The callback ciphertext is a variant of the sealed-box construction,
     extended with negotiated-suite selection and cleartext-header AAD
     binding -- see [seal_local_rp_callback] / [open_local_rp_callback].

   This module performs no I/O and never reads the system clock -- every
   "current time" is an explicit [now] parameter, so verification stays
   deterministic and testable against fixed conformance vectors.

   Only the subset actually used by an RP (build+sign the
   descriptor/login-request/ticket-redemption; verify+open the callback) is
   exercised at runtime by this SDK. [build_local_rp_callback_payload] /
   [sign_local_rp_callback_payload] / [seal_local_rp_callback] are IDP-side
   operations -- included here (mirroring [liblinkkeys::local_rp], which
   serves both sides) purely so this package's own test suite can act as a
   self-contained fake IDP in the flow tests. *)

let ctx_local_rp_descriptor = "linkkeys-local-rp-descriptor"
let ctx_local_rp_login_request = "linkkeys-local-rp-login-request"
let ctx_local_rp_callback = "linkkeys-local-rp-callback"
let ctx_local_rp_ticket_redemption = "linkkeys-local-rp-ticket-redemption"
let default_clock_skew_seconds = 300.0
let local_rp_callback_box_tag = "linkkeys-local-rp-callback-box"

(* [CBOR([context, payload_bytes])] -- a two-element CBOR array, context
   string first (CBOR text string), then the exact payload bytes (CBOR byte
   string). Deliberately NOT a bare [context || payload] concatenation. *)
let envelope_signature_input (context : string) (payload_bytes : string) : string =
  Cbor.encode (Cbor.Array [ Cbor.Text context; Cbor.Bytes payload_bytes ])

(* ------------------------------------------------------------------ *)
(* Timestamps / expirations                                            *)
(* ------------------------------------------------------------------ *)

(* Check an (issued_at, expires_at) pair against [now], tolerant of
   [skew_seconds] of clock skew in either direction. Boundaries are
   inclusive: exactly [now - skew == expires_at] still passes, and exactly
   one second past either boundary fails. *)
let check_timestamps (issued_at : string) (expires_at : string) (now : float) (skew_seconds : float) : unit =
  let issued, expires =
    try (Timeutil.parse_rfc3339 issued_at, Timeutil.parse_rfc3339 expires_at)
    with Timeutil.Bad_timestamp msg -> Error.raise_ (Error.Bad_timestamp msg)
  in
  if now +. skew_seconds < issued then Error.raise_ Error.Not_yet_valid;
  if now -. skew_seconds > expires then Error.raise_ Error.Expired

(* Named [Level_*] throughout (rather than the shorter [Ok]/[Expired]) to
   avoid shadowing the stdlib [Result.Ok] constructor and this module's own
   [Error] exception helper inside this file's scope. *)
type expiration_level =
  | Level_ok
  | Level_notice
  | Level_warning
  | Level_critical
  | Level_expired

type expiration_status = { level : expiration_level; expires_at : float; now : float }

(* [check_expirations(expires_at, now) -> ExpirationStatus] (design doc,
   "Expiration Helper"): notice at 180 days remaining, warning at 90,
   critical at 30, expired once now >= expires_at. No clock-skew tolerance
   (unlike [check_timestamps]) -- expiry warnings are advisory,
   day-granularity facts, not a replay/freshness security boundary. *)
let check_expirations (expires_at : string) (now : float) : expiration_status =
  let expires = try Timeutil.parse_rfc3339 expires_at with Timeutil.Bad_timestamp msg -> Error.raise_ (Error.Bad_timestamp msg) in
  let remaining = expires -. now in
  let day = 86400.0 in
  let level =
    if now >= expires then Level_expired
    else if remaining <= 30.0 *. day then Level_critical
    else if remaining <= 90.0 *. day then Level_warning
    else if remaining <= 180.0 *. day then Level_notice
    else Level_ok
  in
  { level; expires_at = expires; now }

(* ------------------------------------------------------------------ *)
(* Nonce/state/audience/issuer/callback-url checks                     *)
(* ------------------------------------------------------------------ *)

(* Constant-time comparison ([Eqaf.equal], not [<>]) -- SEC fix: nonce and
   state are unpredictable-to-the-attacker secrets the app committed to at
   [begin_local_login] time, so a timing side channel on the comparison
   must not leak how many leading bytes an attacker-supplied value got
   right. Audience/issuer/callback-URL below are NOT comparably
   attacker-guessable secrets (they're public-ish routing facts, not
   unpredictable tokens the app is relying on as proof of possession), so
   they intentionally keep ordinary structural equality. *)
let verify_nonce_state (expected_nonce : string) (expected_state : string) (actual_nonce : string) (actual_state : string) : unit =
  if not (Eqaf.equal expected_nonce actual_nonce) then Error.raise_ Error.Nonce_mismatch;
  if not (Eqaf.equal expected_state actual_state) then Error.raise_ Error.State_mismatch

let verify_audience (payload_audience_fingerprint : string) (local_rp_fingerprint : string) : unit =
  if payload_audience_fingerprint <> local_rp_fingerprint then Error.raise_ Error.Audience_mismatch

let verify_issuer (payload_user_domain : string) (expected_domain : string) : unit =
  if payload_user_domain <> expected_domain then Error.raise_ Error.Issuer_mismatch

let verify_callback_url (payload_callback_url : string) (arrived_url : string) : unit =
  if payload_callback_url <> arrived_url then Error.raise_ Error.Callback_url_mismatch

(* ------------------------------------------------------------------ *)
(* Descriptor (build + sign only -- verification is the IDP's job)     *)
(* ------------------------------------------------------------------ *)

(* [fingerprint] is always derived from [signing_public_key] -- callers
   cannot set it directly, so it can never drift from the key it names. *)
let build_local_rp_descriptor ~app_name ~local_domain_hint ~signing_public_key ~encryption_public_key ~supported_suites
    ~created_at ~expires_at : Types.Local_rp_descriptor.t =
  {
    app_name;
    local_domain_hint;
    signing_public_key;
    encryption_public_key;
    fingerprint = Crypto.fingerprint signing_public_key;
    supported_suites;
    created_at;
    expires_at;
  }

let sign_local_rp_descriptor (descriptor : Types.Local_rp_descriptor.t) (private_key : string) :
    Types.Signed_local_rp_descriptor.t =
  let descriptor_bytes = Types.Local_rp_descriptor.to_cbor descriptor in
  let signature_input = envelope_signature_input ctx_local_rp_descriptor descriptor_bytes in
  let signature = Crypto.sign_ed25519 private_key signature_input in
  { descriptor = descriptor_bytes; signature }

(* ------------------------------------------------------------------ *)
(* Login request (build + sign only)                                   *)
(* ------------------------------------------------------------------ *)

let build_local_rp_login_request ~descriptor ~callback_url ~nonce ~state ~requested_claims ~required_claims ~issued_at
    ~expires_at : Types.Local_rp_login_request.t =
  { descriptor; callback_url; nonce; state; requested_claims; required_claims; issued_at; expires_at }

let sign_local_rp_login_request (request : Types.Local_rp_login_request.t) (private_key : string) :
    Types.Signed_local_rp_login_request.t =
  let request_bytes = Types.Local_rp_login_request.to_cbor request in
  let signature_input = envelope_signature_input ctx_local_rp_login_request request_bytes in
  let signature = Crypto.sign_ed25519 private_key signature_input in
  { request = request_bytes; signature }

(* ------------------------------------------------------------------ *)
(* Ticket redemption (build + sign -- the RP's possession proof)       *)
(* ------------------------------------------------------------------ *)

let build_local_rp_ticket_redemption_request ~claim_ticket ~fingerprint ~issued_at :
    Types.Local_rp_ticket_redemption_request.t =
  { claim_ticket; fingerprint; issued_at }

let sign_local_rp_ticket_redemption_request (request : Types.Local_rp_ticket_redemption_request.t) (private_key : string) :
    Types.Signed_local_rp_ticket_redemption_request.t =
  let request_bytes = Types.Local_rp_ticket_redemption_request.to_cbor request in
  let signature_input = envelope_signature_input ctx_local_rp_ticket_redemption request_bytes in
  let signature = Crypto.sign_ed25519 private_key signature_input in
  { request = request_bytes; signature }

(* ------------------------------------------------------------------ *)
(* Callback payload (build + sign -- IDP-side, used only by this       *)
(* package's own fake-IDP flow tests) / verify (RP-side, used by       *)
(* Complete_login.complete_local_login)                                *)
(* ------------------------------------------------------------------ *)

let build_local_rp_callback_payload ~user_id ~user_domain ~claim_ticket ~audience_fingerprint ~callback_url ~nonce
    ~state ~issued_at ~expires_at : Types.Local_rp_callback_payload.t =
  { user_id; user_domain; claim_ticket; audience_fingerprint; callback_url; nonce; state; issued_at; expires_at }

let sign_local_rp_callback_payload (payload : Types.Local_rp_callback_payload.t) ~(key_id : string) ~(algorithm : string)
    (private_key : string) : Types.Signed_local_rp_callback_payload.t =
  let payload_bytes = Types.Local_rp_callback_payload.to_cbor payload in
  let signature_input = envelope_signature_input ctx_local_rp_callback payload_bytes in
  let signature =
    try Crypto.sign_with_algorithm algorithm signature_input private_key
    with Crypto.Unsupported_algorithm a -> Error.raise_ (Error.Unsupported_algorithm a)
  in
  { payload = payload_bytes; signing_key_id = key_id; signature }

(* Reject a signing key that is not currently a signing key, or is
   revoked/expired -- shared by every verify path that resolves a key by
   id. *)
let check_signing_key_valid (key : Types.Domain_public_key.t) (now : float) : unit =
  if key.key_usage <> "sign" then Error.raise_ (Error.Signature_invalid "key is not a signing key");
  match Crypto.signing_key_validity key.expires_at key.revoked_at now with
  | Crypto.Revoked -> Error.raise_ (Error.Key_revoked key.key_id)
  | Crypto.Expired | Crypto.Bad_expiry -> Error.raise_ (Error.Key_expired key.key_id)
  | Crypto.Valid -> ()

(* Verify a domain-signed callback payload envelope against a set of domain
   public keys: resolve signing_key_id, reject a
   revoked/expired/non-signing key, verify the envelope signature, decode,
   then check issued_at/expires_at bounds. Nothing inside the payload is
   trusted before this succeeds. *)
let verify_local_rp_callback_payload (signed : Types.Signed_local_rp_callback_payload.t)
    (domain_public_keys : Types.Domain_public_key.t list) (now : float) (skew_seconds : float) :
    Types.Local_rp_callback_payload.t =
  let key =
    match List.find_opt (fun (k : Types.Domain_public_key.t) -> k.key_id = signed.signing_key_id) domain_public_keys with
    | Some k -> k
    | None -> Error.raise_ (Error.Key_not_found signed.signing_key_id)
  in
  check_signing_key_valid key now;
  let signature_input = envelope_signature_input ctx_local_rp_callback signed.payload in
  (try Crypto.resolve_and_verify key.algorithm signature_input signed.signature key.public_key with
  | Crypto.Unsupported_algorithm a -> Error.raise_ (Error.Unsupported_algorithm a)
  | Crypto.Verification_failed _ -> Error.raise_ (Error.Signature_invalid "callback payload signature verification failed"));
  let payload =
    try Types.Local_rp_callback_payload.of_cbor signed.payload
    with Cbor.Decode_error msg -> Error.raise_ (Error.Decode_failed (Printf.sprintf "callback payload: %s" msg))
  in
  check_timestamps payload.issued_at payload.expires_at now skew_seconds;
  payload

(* Cross-check the cleartext callback header's routing fields against the
   authoritative copies inside the decrypted, signature-verified payload.
   The header is already bound as AEAD associated data, but a verifier must
   still consult the signed copies rather than trusting the header alone. *)
let check_callback_header_matches_payload (header : Types.Local_rp_callback_header.t)
    (payload : Types.Local_rp_callback_payload.t) : unit =
  if header.fingerprint <> payload.audience_fingerprint then Error.raise_ (Error.Header_payload_mismatch "fingerprint");
  if header.nonce <> payload.nonce then Error.raise_ (Error.Header_payload_mismatch "nonce");
  if header.state <> payload.state then Error.raise_ (Error.Header_payload_mismatch "state");
  if header.issued_at <> payload.issued_at then Error.raise_ (Error.Header_payload_mismatch "issued_at");
  if header.expires_at <> payload.expires_at then Error.raise_ (Error.Header_payload_mismatch "expires_at")

(* ------------------------------------------------------------------ *)
(* Callback sealed box (Wire Precision: "Callback sealed box")         *)
(* ------------------------------------------------------------------ *)

(* Derive the AEAD key and construct the KDF info/AAD-prefix context:
   [tag || suite_id_utf8 || ephemeral_public(32) || recipient_public(32)],
   raw concatenation. Returns (aead_key, context). *)
let local_rp_callback_kdf (suite : Crypto.AeadSuite.t) (ephemeral_public : string) (recipient_public : string)
    (shared_secret : string) : string * string =
  let suite_id = Crypto.AeadSuite.to_str suite in
  let context = local_rp_callback_box_tag ^ suite_id ^ ephemeral_public ^ recipient_public in
  let key = Crypto.hkdf_sha256_expand shared_secret context 32 in
  (key, context)

(* Seal a SignedLocalRpCallbackPayload into a LocalRpEncryptedCallback for
   [recipient_encryption_public_key], under [suite]. IDP-side operation --
   included here purely so this package's own tests can build a
   self-contained fake IDP (see module docs).

   [?ephemeral_private_key]/[?aead_nonce] are deterministic-testing hooks:
   production callers must leave both [None] so real OS randomness is
   used. *)
let seal_local_rp_callback ?ephemeral_private_key ?aead_nonce (signed_payload : Types.Signed_local_rp_callback_payload.t)
    (suite : Crypto.AeadSuite.t) (recipient_encryption_public_key : string) ~fingerprint ~nonce ~state ~issued_at
    ~expires_at : Types.Local_rp_encrypted_callback.t =
  let ephemeral_private = match ephemeral_private_key with Some k -> k | None -> Crypto.random_bytes 32 in
  let nonce_bytes = match aead_nonce with Some n -> n | None -> Crypto.random_bytes 12 in
  let ephemeral_public = Crypto.x25519_public_from_private ephemeral_private in
  let shared_secret = Crypto.x25519_diffie_hellman ephemeral_private recipient_encryption_public_key in
  Crypto.reject_low_order shared_secret;
  let plaintext = Types.Signed_local_rp_callback_payload.to_cbor signed_payload in
  let header : Types.Local_rp_callback_header.t =
    {
      fingerprint;
      nonce;
      state;
      suite = Crypto.AeadSuite.to_str suite;
      ephemeral_public_key = ephemeral_public;
      aead_nonce = nonce_bytes;
      issued_at;
      expires_at;
    }
  in
  let header_bytes = Types.Local_rp_callback_header.to_cbor header in
  let aead_key, kdf_context = local_rp_callback_kdf suite ephemeral_public recipient_encryption_public_key shared_secret in
  let aad = kdf_context ^ header_bytes in
  let ciphertext = Crypto.aead_encrypt suite aead_key nonce_bytes aad plaintext in
  { header = header_bytes; ciphertext }

(* Open a LocalRpEncryptedCallback with the local RP's encryption private
   key. [allowed_suites] is the local RP's own supported-suite list (from
   its descriptor): a header advertising a suite NOT in that list is
   rejected even if it is otherwise a valid registry id (Wire Precision:
   "The SDK must decrypt only with a suite listed in its own descriptor").

   Returns (header, signed_payload) -- the still-domain-signature-
   unverified payload envelope. Callers must still call
   [verify_local_rp_callback_payload] against fetched domain keys, and then
   [check_callback_header_matches_payload], before trusting the result. *)
let open_local_rp_callback (encrypted : Types.Local_rp_encrypted_callback.t) (recipient_encryption_private_key : string)
    (allowed_suites : Crypto.AeadSuite.t list) : Types.Local_rp_callback_header.t * Types.Signed_local_rp_callback_payload.t =
  let header =
    try Types.Local_rp_callback_header.of_cbor encrypted.header
    with Cbor.Decode_error msg -> Error.raise_ (Error.Decode_failed (Printf.sprintf "callback header: %s" msg))
  in
  let suite =
    match Crypto.AeadSuite.parse_str header.suite with
    | None -> Error.raise_ (Error.Unsupported_suite header.suite)
    | Some s -> s
  in
  if not (List.mem suite allowed_suites) then Error.raise_ (Error.Suite_not_advertised header.suite);
  if String.length header.ephemeral_public_key <> 32 then
    Error.raise_ (Error.Invalid_key_length "ephemeral_public_key must be 32 bytes");
  if String.length header.aead_nonce <> 12 then Error.raise_ (Error.Invalid_key_length "aead_nonce must be 12 bytes");
  let recipient_public = Crypto.x25519_public_from_private recipient_encryption_private_key in
  let shared_secret =
    try Crypto.x25519_diffie_hellman recipient_encryption_private_key header.ephemeral_public_key
    with Crypto.Error msg -> Error.raise_ (Error.Decryption_failed (Printf.sprintf "callback decryption failed: %s" msg))
  in
  Crypto.reject_low_order shared_secret;
  let aead_key, kdf_context = local_rp_callback_kdf suite header.ephemeral_public_key recipient_public shared_secret in
  let aad = kdf_context ^ encrypted.header in
  let plaintext =
    try Crypto.aead_decrypt suite aead_key header.aead_nonce aad encrypted.ciphertext
    with Crypto.Error msg -> Error.raise_ (Error.Decryption_failed (Printf.sprintf "callback decryption failed: %s" msg))
  in
  let signed_payload =
    try Types.Signed_local_rp_callback_payload.of_cbor plaintext
    with Cbor.Decode_error msg -> Error.raise_ (Error.Decode_failed (Printf.sprintf "callback payload: %s" msg))
  in
  (header, signed_payload)
